pkg/registry/core/service/storage/rest.go

```diff
@@ -541,33 +572,316 @@ func (r *REST) ConvertToTable(ctx context.Context, object runtime.Object, tableO
+	if oldService.Spec.Type == api.ServiceTypeExternalName && service.Spec.Type != api.ServiceTypeExternalName {
+		allocated, err := rs.allocServiceClusterIPs(service)
+		return allocated, nil, err
+	}
+
+	// CASE B:
+	// Update service from non-ExternalName to ExternalName, should release ClusterIP if exists.
+	if oldService.Spec.Type != api.ServiceTypeExternalName && service.Spec.Type == api.ServiceTypeExternalName {
+		released, err := rs.releaseServiceClusterIPs(oldService)
+		return nil, released, err
+	}
+
+	// if headless service then we bail out early (no clusterIPs management needed)
+	if len(oldService.Spec.ClusterIPs) > 0 && oldService.Spec.ClusterIPs[0] == api.ClusterIPNone {
+		return nil, nil, nil
+	}
+
+	upgraded := len(oldService.Spec.IPFamilies) == 1 && len(service.Spec.IPFamilies) == 2
+	downgraded := len(oldService.Spec.IPFamilies) == 2 && len(service.Spec.IPFamilies) == 1
+
+	// CASE C:
+	if upgraded {
+		toAllocate := make(map[api.IPFamily]string)
+		// if secondary ip was named, just get it. if not add a marker
+		if len(service.Spec.ClusterIPs) < 2 {
+			service.Spec.ClusterIPs = append(service.Spec.ClusterIPs, "" /* marker */)
+		}
+
+		toAllocate[service.Spec.IPFamilies[1]] = service.Spec.ClusterIPs[1]
+
+		// allocate
+		allocated, err := rs.allocClusterIPs(service, toAllocate)
+		// set if successful
+		if err == nil {
+			service.Spec.ClusterIPs[1] = allocated[service.Spec.IPFamilies[1]]
+		}
+
+		return allocated, nil, err
+	}
+
+	// CASE D:
+	if downgraded {
+		toRelease := make(map[api.IPFamily]string)
+		toRelease[oldService.Spec.IPFamilies[1]] = oldService.Spec.ClusterIPs[1]
+		released, err := rs.releaseClusterIPs(toRelease)
+		return nil, released, err
+	}
+	// it was not an upgrade nor downgrade
+	return nil, nil, nil
+}
+
+// releases allocated ClusterIPs for service that is about to be deleted
+func (rs *REST) releaseServiceClusterIPs(service *api.Service) (released map[api.IPFamily]string, err error) {
+	// external name don't get ClusterIPs
+	if service.Spec.Type == api.ServiceTypeExternalName {
+		return nil, nil
+	}
+
+	// headless don't get ClusterIPs
+	if len(service.Spec.ClusterIPs) > 0 && service.Spec.ClusterIPs[0] == api.ClusterIPNone {
+		return nil, nil
+	}
+
+	toRelease := make(map[api.IPFamily]string)
+	for i, ipFamily := range service.Spec.IPFamilies {
+		// during release we depend on the fact that on create || update path
+		// families + clusterIPs were set correctly
+		toRelease[ipFamily] = service.Spec.ClusterIPs[i]
+	}
+	return rs.releaseClusterIPs(toRelease)
+}
+
+// attempts to default service ip families according to cluster configuration
+// while ensuring that provided families are configured on cluster.
+func (rs *REST) tryDefaultValidateServiceClusterIPFields(service *api.Service) error {
+	// can not do anything here
+	if service.Spec.Type == api.ServiceTypeExternalName {
+		return nil
+	}
+	// two families or two IPs with Single or PreferDualStack is not allowed
```

Comment:
just asking, so if I PreferDualStack I can only set on family, I can not ask for 2

Reply:
Yes. PreferDualstack is `give me two families if you have them.` But if user set two families then they are asking for `dualstack` as in `RequireDualstack`